refactoring goto statements

yujiex opened the pull request:
Pull request overview
---------------------
 - This PR aims to convert the goto statements in the physics-based VRF model to use more standard while or do-while loops.

## pattern 1

The following table shows some pseudo-code of the current goto example in the physics-based VRF model (column 1) and standard while and do-while loops (columns 2 and 3). To simplify the reasoning of its execution, I simplified the operations into letters and will check what type of regex they each produce. 

![image](https://github.com/NREL/EnergyPlus/assets/13421161/45b6ff1a-3309-4289-87c6-2b59227f5a81)

A = stuff A, B = stuff B and counter increment, C being the condition check

For this pattern of existing code. I'll try to hoist the "stuff B" part out of the if branch and have an intermediate step

```
    label:
        stuff A
        evaluate condition (in case stuff overwrites variables used in it)
        stuff B
    if (condition) {
        goto label
    }
```

Then this can be converted to a do-while loop

```
    do {
        stuff A
        evaluate condition (in case stuff overwrites variables used in it)
        stuff B
    } while (condition)
```

suppose D = evaluate condition (in case stuff overwrites variables used in it), then both the intermediate and final step evaluates to (ADBC)+

If A and B are the things that matter most, than the original version of code will execute to A(BA)*, and the new code will execute to (AB)+, with a trailing B comparing with the original code. If this additional B doesn't impact the result, then this could be a way to transition such a pattern of goto to a standard while loop.

Otherwise, the following should be used, which evaluates to ADC(BCADC)\*. If D and C doesn't change the behavior, then it evaluates to A(BA)*

```
    do {
        stuff A
        evaluate condition (in case stuff overwrites variables used in it)
        if (condition) {
            stuff B
        }
    } while (condition)
```

## pattern 2

goto jumps out two or more levels, like in label23 in `HVACVariableRefrigerantFlow.cc`.

```
Label:;
stuff A
if (C) {
    stuff E
} else {
    stuff F
    if (D) {
        stuff B
        goto label
    }
    stuff G
}
```
This can be converted to a do-while loop as follows
```
do {
    stuff A
    if (C) {
        stuff E
        break
    } else {
        stuff F
        if (D) {
            stuff B
        } else {
            stuff G
            break
        }
    }
} while (D)
```
In both cases, the following is what it evaluates to for different combination of C and D
```
C true  D true:  ACE
C true  D false: ACE
C false D true:  (ACFDB)+ACFDG
C false D false: ACFDG
```

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Commit:
goto to do-while for Label20 in physics-based VRF

original:

    label:
        stuff A
    if (some condition) {
        stuff B
        goto label
    }

intermediate step:

    label:
        stuff A
        calculate condition before updating Ncomp
        stuff B
    if (some condition) {
        goto label
    }

final step:

    do {
        stuff A
        calculate condition before updating Ncomp
        stuff B
    } while (condition)

Commit:
goto to do-while for Label20 in physics-based VRF

original:

    label:
        stuff A
    if (some condition) {
        stuff B
        goto label
    }

intermediate step:

    label:
        stuff A
        calculate condition before updating Ncomp
        stuff B
    if (some condition) {
        goto label
    }

final step:

    do {
        stuff A
        stuff B
    } while (condition)

Commit:
Merge branch 'refactorGOTO' of https://github.com/NREL/EnergyPlus into refactorGOTO

amirroth commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -11687,36 +11687,36 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state)
 
             // Iteration_Ncomp: Perform iterations to calculate Ncomp (Label20)
             Counter = 1;
-        Label20:;
```
🙏

Myoldmopar commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -11687,36 +11687,36 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state)
+                                      CompSpdActual,
+                                      Ncomp_new);
+
+                bool not_converged = (std::abs(Ncomp_new - Ncomp) > (Tolerance * Ncomp));
```
If you are still making changes in this PR, could I make the tiniest of requests?  Could you make the variable name `converged` instead of `not_converged`?  If we ever need to check if we are converged, we have to check `!not_converged`, and that's a more difficult mental parse, at least to me.  If you are done touching this code, then feel free to ignore, it's not all that important :)

yujiex replied:
Thanks for the suggestion @Myoldmopar. I will change it to `converged` instead. I will continue working on this.

Commit:
Use "converged" instead of "not_converged"

Commit:
Merge remote-tracking branch 'origin/develop' into refactorGOTO

Commit:
add another check to remove the one additional Ncomp = Ncomp_new;

assume the evaluation of conditions doesn't change code behavior

this evaluates to A(BA)*

    label:             <- the original code
        stuff A
    if (condition) {
        stuff B
        goto label
    }

this evaluates to (AB)+

    do {
        stuff A
        evaluate condition (in case stuff overwrites variables used in it)
        stuff B
    } while (condition)

This will lead to one more "B" than the original code

So change it to this

    do {
        stuff A
        evaluate condition (in case stuff overwrites variables used in it)
        if (condition) {
            stuff B
        }
    } while (condition)

Commit:
add suffix to "converged", preventing collision with others

Commit:
convert goto label10, same pattern as label20

Commit:
convert label11, this can directyly convert to do-while

evaluated diffs from VariableRefrigerantFlow_FluidTCtrl_HR_5Zone.idf

The following seems to have the same output

    label:
        stuff A
        evaluate condition (in case stuff overwrites variables used in it)
    if (condition) {
        stuff B
        goto label
    }

vs

    label:
        stuff A
        evaluate condition (in case stuff overwrites variables used in it)
        stuff B
    if (condition) {
        goto label
    }

Commit:
convert label 19, can directly convert to do-while

there's only a counter increase inside the if condition in goto

Commit:
convert another label 11, directly convert to do-while

Commit:
convert label13, direct conversion to do-while

Commit:
convert label12, direct to do-while

Commit:
convert label230, direct conversion to do-while

Commit:
Merge remote-tracking branch 'origin/develop' into refactorGOTO

nrel-bot-2c commented:
@yujiex @myoldmopar it has been 28 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex @myoldmopar it has been 28 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex @myoldmopar it has been 28 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 20 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 8 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 8 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

Commit:
convert label 23 to do-while loop

Commit:
Revert "convert label 23 to do-while loop"

This reverts commit f5de0da6bea174afc7bb3c90725ec38e279a0780.

Commit:
Revert "convert label230, direct conversion to do-while"

This reverts commit 069595ab285b53892e25715dc6bd467031d8402d.

Commit:
Revert "convert label12, direct to do-while"

This reverts commit fa739c73941cc0490dde0a52338d7abdcc608104.

Commit:
Revert "convert label13, direct conversion to do-while"

This reverts commit 1bc91cad6291e8c8844d5c12a5a43686d5c1a6de.

Commit:
Revert "convert another label 11, directly convert to do-while"

This reverts commit 46c89c5451527fa8a1d5cbf2bd0b427f494bdd6a.

Commit:
Revert "convert label 19, can directly convert to do-while"

This reverts commit de2818754703d35ad6c6798286d92c50a3395d64.

Commit:
Revert "convert label11, this can directyly convert to do-while"

This reverts commit 76edf3a3b43e3a988946e270d3cb0a6fdc2c3503.

Commit:
Revert "convert goto label10, same pattern as label20"

This reverts commit 740ae7168e1225fc413c4b99e59be1bcc7617400.

Commit:
merge develop

Commit:
apply back the simpler conversion to do-while loop

Commit:
convert label23 to do-while

Myoldmopar commented:
This is great.  One wish is that while in here, the 1300 line function for CalcVRF or whatever it's called could have been split into a few meaningful functions.  But that's fine.  I'm not sure why the rest of CI didn't run here, but I suspect when the curve API branch merges, there could be a conflict or two to resolve.  I'll tackle it and let you know if I have any issues.

Myoldmopar commented:
Wow, no conflicts.  OK, I'm still gonna pull in the latest develop for a quick sanity check.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into refactorGOTO
